src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -263,73 +366,137 @@ public String getRemote() {
     private void doRetrieve(SCMHeadObserver observer, TaskListener listener, GHRepository repo) throws IOException, InterruptedException {
         SCMSourceCriteria criteria = getCriteria();
 
-        listener.getLogger().format("%n  Getting remote branches...%n");
-        int branches = 0;
-        for (Map.Entry<String,GHBranch> entry : repo.getBranches().entrySet()) {
-            final String branchName = entry.getKey();
-            if (isExcluded(branchName)) {
-                continue;
-            }
-            listener.getLogger().format("%n    Checking branch %s%n", HyperlinkNote.encodeTo(repo.getHtmlUrl().toString() + "/tree/" + branchName, branchName));
-            if (criteria != null) {
-                SCMSourceCriteria.Probe probe = getProbe(branchName, "branch", "refs/heads/" + branchName, repo, listener);
-                if (criteria.isHead(probe, listener)) {
-                    listener.getLogger().format("    Met criteria%n");
-                } else {
-                    listener.getLogger().format("    Does not meet criteria%n");
+        Set<String> originBranchesWithPR = new HashSet<>();
```

Comment:
To implement `buildOriginBranch && !buildOriginBranchWithPR` we need to first find the pull requests, so we can skip corresponding origin branches later. Awkward.